packages/flutter_tools/test/asset_bundle_package_test.dart

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:file/file.dart';
+import 'package:file/memory.dart';
+
+import 'package:flutter_tools/src/asset.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/cache.dart';
+
+import 'package:test/test.dart';
+
+import 'src/common.dart';
+import 'src/context.dart';
+
+void main() {
+  void writePubspecFile(String path, String name, {List<String> assets}) {
+    String assetsSection;
+    if (assets == null) {
+      assetsSection = '';
+    } else {
+      final StringBuffer buffer = new StringBuffer();
+      buffer.write(
+'''
+flutter:
+     assets:
+'''
+      );
+
+      for (String asset in assets) {
+        buffer.write(
+'''
+       - $asset
+''')
+        ;
+      }
+      assetsSection = buffer.toString();
+    }
+
+    fs.file(path)
+      ..createSync(recursive: true)
+      ..writeAsStringSync(
+'''
+name: $name
+dependencies:
+  flutter:
+    sdk: flutter
+$assetsSection
+'''
+      );
+  }
+
+  void writePackagesFile(String packages) {
+    fs.file(".packages")
+      ..createSync()
+      ..writeAsStringSync(packages);
+  }
+
+  Future<Null> buildAndVerifyAssets(
+    List<String> assets,
+    List<String> packages,
+    String expectedAssetManifest,
+  ) async {
+    final AssetBundle bundle = new AssetBundle();
+    await bundle.build(manifestPath: 'pubspec.yaml');
+
+    for (String packageName in packages) {
+      for (String asset in assets) {
+        final String entryKey = 'packages/$packageName/$asset';
+        expect(bundle.entries.containsKey(entryKey), true);
+        expect(
+          UTF8.decode(await bundle.entries[entryKey].contentsAsBytes()),
+          asset,
+        );
+      }
+    }
+
+    expect(
+      UTF8.decode(await bundle.entries['AssetManifest.json'].contentsAsBytes()),
+      expectedAssetManifest,
+    );
+  }
+
+  group('AssetBundle assets from package', () {
+    testUsingContext('One package with no assets', () async {
+      // Setting flutterRoot here so that it picks up the MemoryFileSystem's
+      // path separator.
+      Cache.flutterRoot = getFlutterRoot();
```

Comment:
Maybe move this to an `establishFlutterRoot` function so you don't have to duplicate the comment in every test.